### PR TITLE
fix: regenerator-runtime and reenable SES (v1.1.0) on iOS (JSC)

### LINKED ISSUE #8015: [Sentry] TypeError: Attempted to assign to readonly property.
Sentry Issue: [METAMASK-MOBILE-25WQ](https://metamask.sentry.io/issues/4699573603/?referrer=github_integration)

```
TypeError: Attempted to assign to readonly property.
  at resolveEnsToIpfsContentId (app/lib/ens-ipfs/resolver.js:26:42)
  at resolveEnsToIpfsContentId (app/lib/ens-ipfs/resolver.js:71:2)
  at <global> (app/lib/ens-ipfs/resolver.js:11:56)
  at handleIpfsContent (app/components/Views/BrowserTab/index.js:470:38)
  at handleIpfsContent (app/components/Views/BrowserTab/index.js:531:6)
...
(45 additional frame(s) were not displayed)
```

### LINKED ISSUE #8017: TypeError: Attempted to assign to readonly property.
Sentry Issue: [METAMASK-MOBILE-25YQ](https://metamask.sentry.io/issues/4702324986/?referrer=github_integration)

```
TypeError: Attempted to assign to readonly property.
  at getQuote (node_modules/@consensys/on-ramp-sdk/dist/providers/Transak/TransakSDK.js:75:9)
  at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:20)
  at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:22:9)
  at tryCallOne (node_modules/promise/setimmediate/core.js:37:12)
  at setImmediate$argument_0 (node_modules/promise/setimmediate/core.js:123:15)
...
(10 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by leotm**
## **Description**

Fix recent spike in iOS crashes (sending ETH on Optimism)
and reenable SES on iOS (we disabled it temporarily for our v7.12.2 hotfix)
_nb: [flakey CI](https://github.com/MetaMask/metamask-mobile/pull/8046) earlier was Node 18.19, not SES_

Make `regenerator-runtime@0.11.1` compatible with SES via a [patch](https://github.com/MetaMask/metamask-mobile/pull/8033/commits/c5eba9da3b9cb9000eb14e7223da81c830e00f5a) (updated since to a resolution)
- refactor generator function prototype iterator symbol
- update object property assignment to `Object.defineProperty`

```console
...
├─┬ @metamask/assets-controllers@7.0.0
│ └─┬ babel-runtime@6.26.0
│   └── regenerator-runtime@0.11.1
...
```

<details><summary>full regenerator-runtime dep tree</summary>

```console
├─┬ @babel/runtime@7.23.2
│ └── regenerator-runtime@0.14.0
├─┬ @metamask/assets-controllers@7.0.0
│ └─┬ babel-runtime@6.26.0
│   └── regenerator-runtime@0.11.1
├─┬ @storybook/addon-knobs@5.3.21
│ └─┬ @storybook/api@5.3.21
│   └── regenerator-runtime@0.13.9 deduped
├─┬ @storybook/addon-ondevice-actions@6.5.6
│ └─┬ @storybook/addons@6.5.16
│   ├─┬ @storybook/api@6.5.16
│   │ └── regenerator-runtime@0.13.11 deduped
│   ├─┬ @storybook/router@6.5.16
│   │ └── regenerator-runtime@0.13.11
│   ├─┬ @storybook/theming@6.5.16
│   │ └── regenerator-runtime@0.13.11 deduped
│   └── regenerator-runtime@0.13.11
├─┬ @storybook/react-native@6.5.6
│ ├─┬ @storybook/client-api@6.5.16
│ │ ├─┬ @storybook/store@6.5.16
│ │ │ └── regenerator-runtime@0.13.11
│ │ └── regenerator-runtime@0.13.11
│ ├─┬ @storybook/core-client@6.5.16
│ │ ├─┬ @storybook/ui@6.5.16
│ │ │ ├─┬ @storybook/api@6.5.16
│ │ │ │ └── regenerator-runtime@0.13.11 deduped
│ │ │ ├─┬ @storybook/components@6.5.16
│ │ │ │ └── regenerator-runtime@0.13.11 deduped
│ │ │ ├─┬ @storybook/theming@6.5.16
│ │ │ │ └── regenerator-runtime@0.13.11 deduped
│ │ │ └── regenerator-runtime@0.13.11
│ │ └── regenerator-runtime@0.13.11
│ └─┬ @storybook/preview-web@6.5.16
│   └── regenerator-runtime@0.13.11
├─┬ appium@1.22.3
│ ├─┬ appium-android-driver@4.54.0
│ │ └─┬ jimp@0.16.2
│ │   ├─┬ @jimp/custom@0.16.2
│ │   │ └─┬ @jimp/core@0.16.2
│ │   │   └─┬ @jimp/utils@0.16.2
│ │   │     └── regenerator-runtime@0.13.11
│ │   └── regenerator-runtime@0.13.11
│ └─┬ appium-tizen-driver@1.1.1-beta.5
│   └─┬ jimp@0.5.6
│     └─┬ @babel/polyfill@7.12.1
│       └── regenerator-runtime@0.13.11
├─┬ react-native@0.71.14 invalid: "^0.0.0-0 || 0.60 - 0.70 || 1000.0.0" from node_modules/@react-native-async-storage/async-storage, "^0.56.0" from node_modules/react-native-jazzicon, "^0.62.2" from node_modules/react-native-v8
│ └── regenerator-runtime@0.13.11
└── regenerator-runtime@0.13.9
```

</details>

Considerations
- further patch `runtime.js` other relevant parts
- or resolve `@metamask/assets-controllers>babel-runtime>regenerator-runtime`
  - from v0.11.1 to v0.13.8+ (containing [fix](https://github.com/facebook/regenerator/commit/903a50726da73d0ffd5b57a7617bbcc2fd4bbc99))
- or resolve `@metamask/assets-controllers>babel-runtime`
  - from v6.26.0 to [v7.8.4](https://github.com/babel/babel/blame/9015fda3c106abda6928cf348797a38f09d4b3e0/packages/babel-runtime/package.json)+ (regenerator-runtime: ^0.13.2, which locks to 0.13.11)
- [ethjs patches](https://github.com/LavaMoat/docs/issues/20) still needed?

_Remaining `regenerator-runtime` transitive deps are safe above v0.13.8 (see dep tree above)_

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/8015

- https://metamask.sentry.io/issues/4699573603/?project=2299799 sev1
  - _app/lib/ens-ipfs/resolver.js26:42_ in `resolveEnsToIpfsContentId`
    - `const Registry = contract(registryAbi).at(registryAddress);` an ethjs/ethjs-contract
  - _Suspect Commit: https://github.com/MetaMask/metamask-mobile/commit/b677fd95504651991fc4ce6624180fa47f155c4d update to `resolveEnsToIpfsContentId`_
- https://metamask.sentry.io/issues/4701946024/?project=2299799

Screenshot 2023-12-07 at 4 03 00 pm

Fixes: https://github.com/MetaMask/metamask-mobile/issues/8017

- https://metamask.sentry.io/issues/4702324986/?project=2299799

Additional context
- https://github.com/MetaMask/metamask-extension/pull/10663
- https://github.com/endojs/endo/discussions/1855

## **Manual testing steps**

Steps in Sentry issue breadcrumbs (iOS prod)

To enable SES in debug-mode, omit `!__DEV__`

https://github.com/MetaMask/metamask-mobile/blob/46c1cea9328ff2a71bbbe98179af755c56385fd3/patches/react-native%2B0.71.14.patch#L11-L12

_I've not managed to repro locally this way yet_

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: patches/babel-runtime++regenerator-runtime+0.11.1.patch [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
++++ b/node_modules/babel-runtime/node_modules/regenerator-runtime/runtime.js
+@@ -414,9 +414,9 @@
+   // iterator prototype chain incorrectly implement this, causing the Generator
+   // object to not be returned from this call. This ensures that doesn't happen.
+   // See https://github.com/facebook/regenerator/issues/274 for more details.
+-  Gp[iteratorSymbol] = function() {
++  Object.defineProperty(Gp, iteratorSymbol, { value: function () {
+     return this;
+-  };
++  } });
```
Comment (leotm) [inline patches/babel-runtime++regenerator-runtime+0.11.1.patch:13]: consider instead

```js
  define(Gp, iteratorSymbol, function() {
    return this;
  });
```

like [we do in mm-ext](https://github.com/MetaMask/metamask-extension/blame/0645fa6a6be31f3813abb040f8fe801afedfa89c/.yarn/patches/regenerator-runtime-npm-0.13.7-41bcbe64ea.patch#L45-L49) if `define` is a global

Screenshot 2023-12-07 at 3 51 22 pm

Comment (leotm) [reply patches/babel-runtime++regenerator-runtime+0.11.1.patch:13]: also consider adding code comment

**Comment by leotm** [comment]
converted back to draft until failing/cancelled checks resolved

**Comment by leotm** [comment]
pr re-readied up, since CI flakey unit tests unrelated to this pr

**Comment by leotm** [comment]
once
- https://github.com/MetaMask/metamask-mobile/pull/8046

is merged, unit tests should be passing again and ready for QA

we _shouldn't_ need to [disable SES](https://github.com/MetaMask/metamask-mobile/pull/8045)

File: patches/babel-runtime++regenerator-runtime+0.11.1.patch (hunk 1 above)
Comment (leotm) [reply patches/babel-runtime++regenerator-runtime+0.11.1.patch]: patch reverted

```console
...
├─┬ @metamask/assets-controllers@7.0.0
│ └─┬ babel-runtime@6.26.0
│   └── regenerator-runtime@0.11.1
...
```

now resolved to ^0.13.8 (regenerator-runtime@0.13.11) containing [fix](https://github.com/facebook/regenerator/commit/903a50726da73d0ffd5b57a7617bbcc2fd4bbc99)

similar can be done now to cleanup metamask-extension:
for https://github.com/MetaMask/metamask-extension/blame/develop/.yarn/patches/regenerator-runtime-npm-0.13.7-41bcbe64ea.patch

**Comment by leotm** [comment]
> E2E test started on Bitrise: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a273882e-3ca2-4d7d-a332-9df00fe1e0c6 You can also kick off another Bitrise E2E smoke test by removing and re-applying the (Run Smoke E2E) label

build_smoke_e2e_ios_android_stage ✅
run_smoke_e2e_ios_android_stage ❌
iOS `Failed to find Detox executable at path: /Users/vagrant/git/node_modules/.bin/detox` :suspect:
Android `Failed to find Detox executable at path: /bitrise/src/node_modules/.bin/detox` :suspect:

local testing

`watchman shutdown-server && watchman watch-del-all && nvm use && export METAMASK_BUILD_TYPE='main' && export METAMASK_ENVIRONMENT='local' && IGNORE_BOXLOGS_DEVELOPMENT="true" FORCE_BUNDLING=true yarn test:e2e:ios:bitrise:build` ✅ 

`watchman shutdown-server && watchman watch-del-all && nvm use && export METAMASK_BUILD_TYPE='main' && export METAMASK_ENVIRONMENT='local' && IGNORE_BOXLOGS_DEVELOPMENT="true" FORCE_BUNDLING=true yarn test:e2e:ios:bitrise:run --testNamePattern="Regression"` results inc

**Comment by leotm** [comment]
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/d36076eb-cdf7-4179-9d94-15ac845bc5a5

File: patches/react-native+0.71.15.patch [hunk 2]
```diff
@@ -8,10 +8,10 @@ index 1379ffd..e59f511 100644
  
 +const Platform = require('../Utilities/Platform');
 +
-+const IS_LOCKDOWN_ENABLED = false; // Temporarily false until CI issues investigated and resolved
++const IS_LOCKDOWN_ENABLED = true;
 +
 +if (IS_LOCKDOWN_ENABLED && Platform.OS === 'ios' && !global?.HermesInternal && !__DEV__) {
-+  require('../../../../ses.cjs'); // ses@0.18.8
```
Comment (leotm) [inline patches/react-native+0.71.15.patch:14]: version is now displayed in [ses.cjs#1](https://github.com/MetaMask/metamask-mobile/pull/8033/files#diff-4a0c1040ff18397abb6a469a945010cf8e5f0160e1aaff29b810a17f5c259a71R1)

File: ses.cjs [hunk 3]
```diff
@@ -1,8 +1,9 @@
+// ses@1.1.0
```
Comment (Cal-L) [inline ses.cjs:1]: Is this a regenerated file?

**Comment by Cal-L** [comment]
> > E2E test started on Bitrise: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a273882e-3ca2-4d7d-a332-9df00fe1e0c6 You can also kick off another Bitrise E2E smoke test by removing and re-applying the (Run Smoke E2E) label
> 
> build_smoke_e2e_ios_android_stage ✅ run_smoke_e2e_ios_android_stage ❌ iOS `Failed to find Detox executable at path: /Users/vagrant/git/node_modules/.bin/detox` :suspect: Android `Failed to find Detox executable at path: /bitrise/src/node_modules/.bin/detox` :suspect:
> 
> local testing
> 
> `watchman shutdown-server && watchman watch-del-all && git clean -fdx && nvm use && export METAMASK_BUILD_TYPE="main" && export METAMASK_ENVIRONMENT="local" && yarn setup`
> 
> `nvm use && export METAMASK_BUILD_TYPE='main' && export METAMASK_ENVIRONMENT='local' && IGNORE_BOXLOGS_DEVELOPMENT="true" FORCE_BUNDLING=true yarn test:e2e:ios:bitrise:build` ✅
> 
> `nvm use && export METAMASK_BUILD_TYPE='main' && export METAMASK_ENVIRONMENT='local' && IGNORE_BOXLOGS_DEVELOPMENT="true" FORCE_BUNDLING=true yarn test:e2e:ios:bitrise:run --testNamePattern="Smoke"` ✅
> 
> Screenshot 2024-01-19 at 12 22 48 pm
> all smoke tests passing locally, update branch and re-run BitRise Smoke tests (`Rebuild unsuccessful Workflows with remote access` unsuccessful)

I believe there were issues with our E2E tests on Bitrise. @cortisiko @chrisleewilcox could you confirm?

**Comment by Cal-L** [comment]
@leotm Are you aware if the latest core controllers already resolve this issue? If not, could you follow up with a PR to core that performs one of the following?

- resolve @metamask/assets-controllers>babel-runtime>regenerator-runtime
from v0.11.1 to v0.13.8+ (containing [fix](https://github.com/facebook/regenerator/commit/903a50726da73d0ffd5b57a7617bbcc2fd4bbc99))
- resolve @metamask/assets-controllers>babel-runtime
from v6.26.0 to [v7.8.4](https://github.com/babel/babel/blame/9015fda3c106abda6928cf348797a38f09d4b3e0/packages/babel-runtime/package.json)+ (regenerator-runtime: ^0.13.2, which locks to 0.13.11)

**Review by Cal-L — COMMENTED** [review]
left a comment and some questions

File: ses.cjs (hunk 3 above)
Comment (leotm) [reply ses.cjs:1]: it's not regenerated but grabbed atm via `curl -O https://npmfs.com/download/ses/1.1.0/dist/ses.cjs`

Comment (leotm) [reply ses.cjs:1]: the original reason for switching to the shim earlier were tricky compat errors with react native
but we could switch to the npm module in future if preferred
then introduce e.g. renovatebot to auto-update SES

**Comment by leotm** [comment]
> @leotm Are you aware if the latest core controllers already resolve this issue? If not, could you follow up with a PR to core that performs one of the following?
> 
> * resolve @metamask/assets-controllers>babel-runtime>regenerator-runtime
>   from v0.11.1 to v0.13.8+ (containing [fix](https://github.com/facebook/regenerator/commit/903a50726da73d0ffd5b57a7617bbcc2fd4bbc99))
> * resolve @metamask/assets-controllers>babel-runtime
>   from v6.26.0 to [v7.8.4](https://github.com/babel/babel/blame/9015fda3c106abda6928cf348797a38f09d4b3e0/packages/babel-runtime/package.json)+ (regenerator-runtime: ^0.13.2, which locks to 0.13.11)

good question after checking [core(main)](https://github.com/MetaMask/core) appears fixed now

```
➜  core git:(main)
@metamask/core-monorepo@109.0.0 /Users/leo/Documents/GitHub/core
└─┬ @metamask/transaction-controller@20.0.0 -> ./packages/transaction-controller
  └─┬ babel-runtime@6.26.0
    └── regenerator-runtime@0.11.1
```

since `@metamask/assets-controllers` no longer relies on `regenerator-runtime`
removed in [`@metamask/core-monorepo@67.0.0`](https://github.com/MetaMask/core/commits/v67.0.0/packages/assets-controllers/package.json) (https://github.com/MetaMask/core/pull/1504), which is`@metamask/assets-controllers` v10
so upgrading our `@metamask/assets-controllers` from v7 to v10-24 is SES-compatible, running local tests on those

our final SES lockdown incompatibility lurks now in `@metamask/transaction-controller`, following up there too

**Comment by Gudahtt** [comment]
I don't think updating that package would help; `babel-runtime` is the dependency of literally dozens of other packages as well, including some core packages.

**Comment by leotm** [comment]
> our final SES lockdown incompatibility now lurks in `@metamask/transaction-controller`
> so following up there with the resolution, unless we can do similar refeactor to remove `babel-runtime` deps?

> I don't think updating that package would help; `babel-runtime` is the dependency of dozens of other packages as well, including some core packages.

yeah i see what you mean

```
@metamask/core-monorepo@109.0.0
└─┬ @metamask/transaction-controller@20.0.0 -> ./packages/transaction-controller
  ├── babel-runtime@6.26.0
  └─┬ eth-method-registry@3.0.0
    └─┬ @metamask/ethjs@0.5.1
      ├─┬ @metamask/ethjs-contract@0.3.4
      │ └── babel-runtime@6.26.0 deduped
      └─┬ @metamask/ethjs-query@0.5.3
        └── babel-runtime@6.26.0 deduped
```

ok leaving it for now, unless a resolution like https://github.com/MetaMask/core/pull/3809 would make sense there

**Review by Cal-L — APPROVED** [review]
LGTM

**Pull request merged**